// @siteimprove/alfa-css/dist/value/color/color.js
import { Parser } from "@siteimprove/alfa-parser";
import { Refinement } from "@siteimprove/alfa-refinement";
import { Selective } from "@siteimprove/alfa-selective";
import { Number, Percentage } from "../numeric/index.js";
import { Keyword } from "../textual/keyword.js";
import { ColorMix } from "./color-mix.js";
import { CSS4Color } from "./css4-color.js";
import { Current } from "./current.js";
import { System } from "./system.js";
const { either } = Parser;
const { or } = Refinement;
/**
 * @public
 */
export var Color;
(function (Color) {
    Color.isCSS4Color = CSS4Color.isCSS4Color;
    Color.isCurrent = Current.isCurrent;
    Color.isSystem = System.isSystem;
    Color.isColorMix = ColorMix.isColorMix;
    function resolve(resolver) {
        return (color) => Selective.of(color)
            .if(Color.isSystem, System.resolve)
            .if(Color.isCurrent, () => resolver.currentColor)
            .else((color) => color.resolve(resolver))
            .get();
    }
    Color.resolve = resolve;
    function partiallyResolve(color) {
        return Selective.of(color)
            .if(Color.isSystem, System.resolve)
            .if(Color.isColorMix, (color) => color.partiallyResolve())
            .else((color) => color)
            .get();
    }
    Color.partiallyResolve = partiallyResolve;
    function toNumber(x, base) {
        const y = x.resolve();
        return Number.isNumber(y) ? y.value / base : y.value;
    }
    /**
     * Creates a color in the sRGB color space.
     */
    function rgb(red, green, blue, alpha = Number.of(1)) {
        return (CSS4Color.of("srgb", [toNumber(red, 255), toNumber(green, 255), toNumber(blue, 255)], toNumber(alpha, 1))
            // We are sure that the color space id exists, so we should always get a
            // correct color.
            .getUnsafe(`Incorrect sRGB color values: ${red}, ${green}, ${blue}, ${alpha}`));
    }
    Color.rgb = rgb;
    /**
     * Creates a color based on its CSS string representation.
     */
    function of(color) {
        return CSS4Color.of(color);
    }
    Color.of = of;
    Color.transparent = CSS4Color.of("transparent").getUnsafe();
    Color.current = Keyword.of("currentcolor");
    Color.system = (value) => Keyword.of(value);
    /**
     * Composite colors of a graphic element ("foreground") over a backdrop
     * ("background").
     * {@link https://drafts.csswg.org/compositing-1/#simplealphacompositing}
     *
     * @remarks
     * The graphic element (foreground) may have a partially transparent color
     * **and** be part of an HTML element which is itself partially transparent.
     * Both transparencies need to be merged first.
     *
     * @param foreground - The color of the graphic element to combine
     * @param background - The color of the backdrop to combine into
     * @param opacity - The opacity of the graphic element, independently of its color.
     */
    function composite(foreground, background, opacity) {
        const foregroundOpacity = foreground.alpha.value * opacity;
        if (foregroundOpacity === 1) {
            return foreground;
        }
        const backgroundOpacity = background.alpha.value * (1 - foregroundOpacity);
        const red = foreground.red.value * foregroundOpacity +
            background.red.value * backgroundOpacity;
        const green = foreground.green.value * foregroundOpacity +
            background.green.value * backgroundOpacity;
        const blue = foreground.blue.value * foregroundOpacity +
            background.blue.value * backgroundOpacity;
        return rgb(Percentage.of(red), Percentage.of(green), Percentage.of(blue), Percentage.of(foregroundOpacity + backgroundOpacity));
    }
    Color.composite = composite;
    /**
     * {@link https://drafts.csswg.org/css-color/#typedef-color}
     */
    function parse(input) {
        return either(Current.parse, System.parse, CSS4Color.parse, ColorMix.parse(parse))(input);
    }
    Color.parse = parse;
    function isTransparent(color) {
        return Selective.of(color)
            .if(Color.isCSS4Color, (color) => color.alpha.value === 0)
            .if(or(Color.isCurrent, Color.isSystem), () => false)
            .else(isTransparentColorMix)
            .get();
    }
    Color.isTransparent = isTransparent;
    function isTransparentColorMix(color) {
        // All the non-transparent components have explicit 0% presence in the mix
        return color.colors.every((item) => isTransparent(item.value) || item.percentage.some((p) => p.value === 0));
    }
})(Color || (Color = {}));
