// @siteimprove/alfa-scraper/dist/screenshot.js
/**
 * @public
 */
export class Screenshot {
    static of(path, type = Screenshot.Type.PNG.of(true)) {
        return new Screenshot(path, type);
    }
    _path;
    _type;
    constructor(path, type) {
        this._path = path;
        this._type = type;
    }
    get path() {
        return this._path;
    }
    get type() {
        return this._type;
    }
    equals(value) {
        return (value instanceof Screenshot &&
            value._path === this._path &&
            value._type.equals(this._type));
    }
    toJSON() {
        return {
            path: this._path,
            type: this._type.toJSON(),
        };
    }
}
/**
 * @public
 */
(function (Screenshot) {
    let Type;
    (function (Type) {
        class PNG {
            static of(background) {
                return new PNG(background);
            }
            _background;
            constructor(background) {
                this._background = background;
            }
            get type() {
                return "png";
            }
            get background() {
                return this._background;
            }
            equals(value) {
                return value instanceof PNG && value._background === this._background;
            }
            toJSON() {
                return {
                    type: "png",
                    background: this._background,
                };
            }
        }
        Type.PNG = PNG;
        class JPEG {
            static of(quality) {
                return new JPEG(quality);
            }
            _quality;
            constructor(quality) {
                this._quality = quality;
            }
            get type() {
                return "jpeg";
            }
            get quality() {
                return this._quality;
            }
            equals(value) {
                return value instanceof JPEG && value._quality === this._quality;
            }
            toJSON() {
                return {
                    type: "jpeg",
                    quality: this._quality,
                };
            }
        }
        Type.JPEG = JPEG;
    })(Type = Screenshot.Type || (Screenshot.Type = {}));
})(Screenshot || (Screenshot = {}));
