// @siteimprove/alfa-rules/dist/sia-r33/rule.js
import { Rule } from "@siteimprove/alfa-act";
import { DOM } from "@siteimprove/alfa-aria";
import { Technique } from "@siteimprove/alfa-wcag";
import { video } from "../common/applicability/video.js";
import { videoTranscript } from "../common/expectation/media-transcript.js";
import { Scope, Stability } from "../tags/index.js";
const { isIncludedInTheAccessibilityTree } = DOM;
export default Rule.Atomic.of({
    uri: "https://alfa.siteimprove.com/rules/sia-r33",
    requirements: [Technique.of("G159")],
    tags: [Scope.Component, Stability.Stable],
    evaluate: ({ device, document }) => {
        return {
            applicability() {
                return video(document, device, { audio: { has: false } });
            },
            expectations(target) {
                return videoTranscript(target, (transcript) => transcript
                    // The transcript element itself might be a container,
                    // e.g. a <span> with accessible text
                    .inclusiveDescendants()
                    .some(isIncludedInTheAccessibilityTree(device)));
            },
        };
    },
});
