// @siteimprove/alfa-selector/dist/selector/pseudo/pseudo-class/has.js
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
import { Cache } from "@siteimprove/alfa-cache";
import { Element, Query } from "@siteimprove/alfa-dom";
import { Iterable } from "@siteimprove/alfa-iterable";
import { Context } from "../../../context.js";
import { Combinator } from "../../index.js";
import { WithSelector } from "./pseudo-class.js";
/**
 * {@link https://drafts.csswg.org/selectors/#has-pseudo}
 */
let Has = (() => {
    var _a;
    let _classSuper = WithSelector;
    let _instanceExtraInitializers = [];
    let __matches_decorators;
    return class Has extends _classSuper {
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            __esDecorate(this, null, __matches_decorators, { kind: "method", name: "_matches", static: false, private: false, access: { has: obj => "_matches" in obj, get: obj => obj._matches }, metadata: _metadata }, null, _instanceExtraInitializers);
            if (_metadata) Object.defineProperty(this, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static of(selector) {
            return new Has(selector);
        }
        constructor(selector) {
            super("has", selector, selector.specificity, selector.useContext);
            __runInitializers(this, _instanceExtraInitializers);
        }
        matches(element, context) {
            return this.useContext
                ? this._matches(element, context ?? Context.empty())
                : // If context is irrelevant, use the empty context to increase cache hit
                    // chances.
                    this._matches(element, Context.empty());
        }
        /**
         * Whether an element match.
         *
         * @remarks
         * This method is further cached because it rebuilds new anchored selectors
         * on each call, so it wouldn't be able to use other caches, notably for
         * ancestors/descendants matching which can be expensive.
         */
        _matches(element, context) {
            const selectors = Iterable.map(this._selector, (selector) => selector.anchoredAt(element));
            return Iterable.some(selectors, (selector) => {
                let candidates;
                // While the relative match could theoretically happen anywhere in the
                // DOM tree, we optimize the search based on the anchor and combinator.
                switch (selector.combinator) {
                    case Combinator.Descendant:
                        candidates = Query.getElementDescendants(element);
                        break;
                    case Combinator.DirectDescendant:
                        candidates = element.children().filter(Element.isElement);
                        break;
                    case Combinator.Sibling:
                        candidates = element
                            .inclusiveSiblings()
                            .skipUntil((elt) => elt === element)
                            .skip(1)
                            .filter(Element.isElement);
                        break;
                    case Combinator.DirectSibling:
                        candidates = element
                            .inclusiveSiblings()
                            .skipUntil((elt) => elt === element)
                            .skip(1)
                            .first()
                            .filter(Element.isElement);
                        break;
                }
                return Iterable.some(candidates, (candidate) => selector.matches(candidate, context));
            });
        }
        *[(__matches_decorators = [(_a = Cache).memoize.bind(_a)], Symbol.iterator)]() {
            yield this;
        }
        equals(value) {
            return value instanceof Has && value._selector.equals(this._selector);
        }
        toJSON() {
            return super.toJSON();
        }
    };
})();
export { Has };
(function (Has) {
    Has.parse = (parseSelector, withColon = true) => WithSelector.parseWithSelector("has", () => parseSelector({ relative: true }), Has.of, withColon);
})(Has || (Has = {}));
