// @siteimprove/alfa-eaa/dist/eaa.js
import { Requirement } from "@siteimprove/alfa-act";
import { None, Option } from "@siteimprove/alfa-option";
import { Criteria } from "./criterion/data.js";
/**
 * @public
 */
export class EAA extends Requirement {
    static of(chapter) {
        return new EAA(chapter);
    }
    _chapter;
    constructor(chapter) {
        super("eaa", "");
        this._chapter = chapter;
    }
    /**
     * The chapter of this criterion.
     */
    get chapter() {
        return this._chapter;
    }
    /**
     * The title of this criterion.
     */
    get title() {
        return Criteria[this._chapter].title;
    }
}
/**
 * @public
 */
(function (EAA) {
    function isChapter(value) {
        return value in Criteria;
    }
    EAA.isChapter = isChapter;
    function isEAA(value) {
        return value instanceof EAA;
    }
    EAA.isEAA = isEAA;
    function fromChapter(chapter) {
        return isChapter(chapter) ? Option.of(EAA.of(chapter)) : None;
    }
    EAA.fromChapter = fromChapter;
})(EAA || (EAA = {}));
