// @siteimprove/alfa-thenable/dist/thenable.js
import { Array } from "@siteimprove/alfa-array";
import { Refinement } from "@siteimprove/alfa-refinement";
const { isFunction, isObject } = Refinement;
/**
 * @public
 */
export var Thenable;
(function (Thenable) {
    /**
     * Check if an unknown value implements the {@link (Thenable:interface)}
     * interface.
     */
    function isThenable(value) {
        return isObject(value) && isFunction(value.then);
    }
    Thenable.isThenable = isThenable;
    function empty() {
        return resolve(undefined);
    }
    Thenable.empty = empty;
    function resolve(value) {
        return defer((resolve) => {
            resolve(value);
        });
    }
    Thenable.resolve = resolve;
    function reject(error) {
        return defer((resolve, reject) => {
            reject(error);
        });
    }
    Thenable.reject = reject;
    function defer(continuation) {
        return new (class Thenable {
            then(resolve, reject) {
                continuation(resolve, reject);
            }
        })();
    }
    Thenable.defer = defer;
    function map(thenable, mapper) {
        return defer((resolved, rejected) => {
            thenable.then((value) => resolved(mapper(value)), rejected);
        });
    }
    Thenable.map = map;
    function apply(thenable, mapper) {
        return flatMap(mapper, (mapper) => map(thenable, mapper));
    }
    Thenable.apply = apply;
    function flatMap(thenable, mapper) {
        return defer((resolved, rejected) => {
            thenable.then((value) => mapper(value).then(resolved, rejected), rejected);
        });
    }
    Thenable.flatMap = flatMap;
    function flatten(thenable) {
        return flatMap(thenable, (thenable) => thenable);
    }
    Thenable.flatten = flatten;
    function all(...thenables) {
        return defer((resolve, reject) => {
            const values = Array.allocate(thenables.length);
            if (thenables.length === 0) {
                return resolve(values);
            }
            let unsettled = thenables.length;
            thenables.forEach((thenables, i) => thenables.then((value) => {
                values[i] = value;
                if (--unsettled === 0) {
                    resolve(values);
                }
            }, (error) => {
                reject(error);
            }));
        });
    }
    Thenable.all = all;
    function any(...thenables) {
        return defer((resolve, reject) => {
            const errors = Array.allocate(thenables.length);
            if (thenables.length === 0) {
                return reject(errors);
            }
            let unsettled = thenables.length;
            thenables.forEach((thenables, i) => thenables.then((value) => {
                resolve(value);
            }, (error) => {
                errors[i] = error;
                if (--unsettled === 0) {
                    reject(errors);
                }
            }));
        });
    }
    Thenable.any = any;
    function race(...thenables) {
        return defer((resolve, reject) => {
            let unsettled = true;
            for (const thenable of thenables) {
                thenable.then((value) => {
                    if (unsettled) {
                        unsettled = false;
                        resolve(value);
                    }
                }, (error) => {
                    if (unsettled) {
                        unsettled = false;
                        reject(error);
                    }
                });
            }
        });
    }
    Thenable.race = race;
})(Thenable || (Thenable = {}));
