// @siteimprove/alfa-promise/dist/promise.js
import { Array } from "@siteimprove/alfa-array";
import * as builtin from "./builtin.js";
/**
 * @public
 */
export var Promise;
(function (Promise) {
    function isPromise(value) {
        return resolve(value) === value;
    }
    Promise.isPromise = isPromise;
    function empty() {
        return resolve(undefined);
    }
    Promise.empty = empty;
    function resolve(value) {
        return builtin.Promise.resolve(value);
    }
    Promise.resolve = resolve;
    function reject(error) {
        return builtin.Promise.reject(error);
    }
    Promise.reject = reject;
    function defer(continuation) {
        return new builtin.Promise(continuation);
    }
    Promise.defer = defer;
    function all(...promises) {
        return defer((resolve, reject) => {
            const values = Array.allocate(promises.length);
            if (promises.length === 0) {
                return resolve(values);
            }
            let unsettled = promises.length;
            promises.forEach((promise, i) => promise.then((value) => {
                values[i] = value;
                if (--unsettled === 0) {
                    resolve(values);
                }
            }, (error) => {
                reject(error);
            }));
        });
    }
    Promise.all = all;
    function any(...promises) {
        return defer((resolve, reject) => {
            const errors = Array.allocate(promises.length);
            if (promises.length === 0) {
                return reject(errors);
            }
            let unsettled = promises.length;
            promises.forEach((promise, i) => promise.then((value) => {
                resolve(value);
            }, (error) => {
                errors[i] = error;
                if (--unsettled === 0) {
                    reject(errors);
                }
            }));
        });
    }
    Promise.any = any;
    function race(...promises) {
        return defer((resolve, reject) => {
            for (const promise of promises) {
                promise.then(resolve, reject);
            }
        });
    }
    Promise.race = race;
})(Promise || (Promise = {}));
