// @siteimprove/alfa-selector/dist/selector/pseudo/pseudo-element/pseudo-element.js
import { Token } from "@siteimprove/alfa-css";
import { Parser } from "@siteimprove/alfa-parser";
const { map, right, take, takeBetween } = Parser;
import { Specificity } from "../../../specificity.js";
import { WithName } from "../../selector.js";
export class PseudoElementSelector extends WithName {
    constructor(name, specificity = Specificity.of(0, 0, 1), useContext = false) {
        super("pseudo-element", name, specificity, useContext);
    }
    equals(value) {
        return value instanceof PseudoElementSelector && super.equals(value);
    }
    toJSON() {
        return {
            ...super.toJSON(),
        };
    }
    toString() {
        return `::${this._name}`;
    }
}
(function (PseudoElementSelector) {
    /**
     * Parses a non-functional, non-legacy pseudo-element (`::<name>`)
     */
    function parseNonLegacy(name, of, withColon = true) {
        return (input) => {
            return map(withColon
                ? right(take(Token.parseColon, 2), Token.parseIdent(name))
                : Token.parseIdent(name), 
            // We need to eta-expand in order to discard the result of parseIdent.
            () => of())(input);
        };
    }
    PseudoElementSelector.parseNonLegacy = parseNonLegacy;
    /**
     * Parses a non-functional, legacy pseudo-element (`::<name>` or `:<name>`)
     */
    function parseLegacy(name, of, withColon = true) {
        return (input) => {
            return map(withColon
                ? right(takeBetween(Token.parseColon, 1, 2), Token.parseIdent(name))
                : Token.parseIdent(name), 
            // We need to eta-expand in order to discard the result of parseIdent.
            () => of())(input);
        };
    }
    PseudoElementSelector.parseLegacy = parseLegacy;
})(PseudoElementSelector || (PseudoElementSelector = {}));
