// @siteimprove/alfa-css/dist/calculation/math-expression/function/clamp.js
import { Real } from "@siteimprove/alfa-math";
import { Parser } from "@siteimprove/alfa-parser";
import { Result } from "@siteimprove/alfa-result";
import { Tuple } from "@siteimprove/alfa-tuple";
import { Function as CSSFunction, Token, } from "../../../syntax/index.js";
import { Unit } from "../../../unit/index.js";
import { Angle, Length, Number } from "../../numeric/index.js";
import { Expression } from "../expression.js";
import { Value } from "../value.js";
import { Function } from "./function.js";
const { isAngle } = Angle;
const { isLength } = Length;
const { isNumber } = Number;
const { isValueExpression } = Value;
const { delimited, mapResult, option, separatedList } = Parser;
/**
 * {@link https://drafts.csswg.org/css-values/#comp-func}
 *
 * @public
 */
export class Clamp extends Function {
    static of(min, value, max) {
        // {@see https://drafts.csswg.org/css-values/#determine-the-type-of-a-calculation}
        const kind = [min, value, max].reduce((old, cur) => old.flatMap((kind) => kind.add(cur.kind)), Result.of(min.kind));
        return kind.map((kind) => new Clamp(min, value, max, kind));
    }
    constructor(min, value, max, kind) {
        super("clamp", [min, value, max], kind);
    }
    get min() {
        return this._args[0];
    }
    get value() {
        return this._args[1];
    }
    get max() {
        return this._args[2];
    }
    reduce(resolver) {
        // We know from the guard in Clamp.of that all args have the same kind.
        const reduced = Tuple.map(this._args, (expr) => expr.reduce(resolver));
        if (Tuple.every(reduced, isValueExpression)) {
            const values = reduced.map((expr) => expr.value);
            const [lower, value, upper] = Tuple.map(values, (value) => value.value);
            // At this point, reduced args should be either:
            // * numbers
            // * angle, in canonical unit (deg)
            // * absolute length, in canonical unit (px)
            // * percentages
            // * relative length, in any unit
            // The first three are reduce-able further; percentages aren't because it
            // may end up being percentages of negative values.
            if (values.every(isNumber)) {
                return Value.of(Number.of(Real.clamp(value, lower, upper)));
            }
            if (Tuple.every(values, 
            // The unit test is theoretically not needed since reduced angle values
            // should always be in the canonical unit (no relative angles)
            (value) => isAngle(value) && value.hasUnit(Unit.Angle.Canonical))) {
                return Value.of(Angle.of(Real.clamp(value, lower, upper), Unit.Angle.Canonical));
            }
            if (Tuple.every(values, (value) => isLength(value) && value.hasUnit(Unit.Length.Canonical))) {
                return Value.of(Length.of(Real.clamp(value, lower, upper), Unit.Length.Canonical));
            }
            // reduced contains percentages or relative lengths, we just fall through
            // to the default case.
        }
        // reduced contains unreduced calculations, we could eagerly compact on the
        // fully reduced ones, but it's easier to just keep everything
        return new Clamp(reduced[0], reduced[1], reduced[2], this._kind);
    }
    toString() {
        return `clamp(${this._args.map((expr) => expr.toString()).join(", ")})`;
    }
}
/** @public */
(function (Clamp) {
    function isClamp(value) {
        return value instanceof Clamp;
    }
    Clamp.isClamp = isClamp;
    Clamp.parse = (parseSum) => mapResult(CSSFunction.parse("clamp", (input) => separatedList(parseSum, delimited(option(Token.parseWhitespace), Token.parseComma), 3, 3)(input)), ([, args]) => Clamp.of(args[0], args[1], args[2]));
})(Clamp || (Clamp = {}));
