// @siteimprove/alfa-css/dist/calculation/math-expression/function/index.js
import { Parser } from "@siteimprove/alfa-parser";
import { Token } from "../../../syntax/index.js";
import { Expression } from "../expression.js";
import { Calculation } from "./calc.js";
import { Clamp } from "./clamp.js";
import { Max } from "./max.js";
import { Min } from "./min.js";
const { exclusive } = Parser;
/** @public */
export var Function;
(function (Function) {
    Function.isCalculation = Calculation.isCalculation;
    Function.isMax = Max.isMax;
    Function.isMin = Min.isMin;
    const names = ["calc", "clamp", "max", "min"];
    const parsers = {
        calc: Calculation.parse,
        clamp: Clamp.parse,
        max: Max.parse,
        min: Min.parse,
    };
    /**
     * Parses a math function in a calculation.
     *
     * @param parseSum - A parser for the sum production in a calculation.
     */
    function parse(parseSum) {
        return exclusive(Token.parseFunction(names), (fn) => parsers[fn.value](parseSum));
    }
    Function.parse = parse;
})(Function || (Function = {}));
